File: backend/app/schemas/attendance.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, field_validator


class CheckinRequest(BaseModel):
    latitude: float = Field(..., ge=-90, le=90, description="User's GPS latitude")
    longitude: float = Field(..., ge=-180, le=180, description="User's GPS longitude")


class CheckoutRequest(BaseModel):
    # Location at checkout is optional (device may not resend GPS)
    latitude: Optional[float] = Field(None, ge=-90, le=90)
    longitude: Optional[float] = Field(None, ge=-180, le=180)


class AttendanceResponse(BaseModel):
    id: int
    user_id: int
    site_id: Optional[int]
    shift_id: Optional[int]
    checkin_time: datetime
    checkout_time: Optional[datetime]
    auto_checkout: bool
    latitude: Optional[float]
    longitude: Optional[float]
    work_duration_minutes: int
    overtime_minutes: int
    is_weekend: bool
    is_holiday: bool
    status: Optional[str]
    created_at: datetime
    site_timezone: str = "Asia/Jakarta"

    model_config = {"from_attributes": True}


class TeamAttendanceRecord(BaseModel):
    id: int
    user_id: int
    employee_id: str
    employee_name: str
    checkin_time: datetime
    checkout_time: Optional[datetime]
    work_duration_minutes: int
    overtime_minutes: int
    is_weekend: bool
    is_holiday: bool
    status: Optional[str]
    created_at: datetime
    site_timezone: str = "Asia/Jakarta"

    model_config = {"from_attributes": True}

    @classmethod
    def from_attendance(cls, att: object) -> "TeamAttendanceRecord":
        return cls(
            id=att.id,
            user_id=att.user_id,
            employee_id=att.user.employee_id,
            employee_name=att.user.name,
            checkin_time=att.checkin_time,
            checkout_time=att.checkout_time,
            work_duration_minutes=att.work_duration_minutes,
            overtime_minutes=att.overtime_minutes,
            is_weekend=att.is_weekend,
            is_holiday=att.is_holiday,
            status=att.status,
            created_at=att.created_at,
            site_timezone=att.site.timezone if att.site else "Asia/Jakarta",
        )


class AutoCheckoutResult(BaseModel):
    processed: int
    message: str

File: web/src/types/attendance.ts
export type AttendanceStatus = 'ONTIME' | 'LATE' | 'OUT_OF_RADIUS'

export interface TeamAttendanceRecord {
  id: number
  user_id: number
  employee_id: string
  employee_name: string
  site_name: string | null
  checkin_time: string        // UTC ISO 8601
  checkout_time: string | null
  work_duration_minutes: number
  overtime_minutes: number
  is_weekend: boolean
  is_holiday: boolean
  status: AttendanceStatus | null
  created_at: string
  site_timezone: string
}

export interface AttendanceDetail {
  id: number
  user_id: number
  site_id: number | null
  shift_id: number | null
  checkin_time: string        // UTC ISO 8601
  checkout_time: string | null
  auto_checkout: boolean
  latitude: number | null
  longitude: number | null
  work_duration_minutes: number
  overtime_minutes: number
  is_weekend: boolean
  is_holiday: boolean
  status: AttendanceStatus | null
  created_at: string
  site_timezone: string
}

export interface AutoCheckoutResult {
  processed: number
  message: string
}
